### PR TITLE
feat: set and retrieve trace id in log macro

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

This pr changes the `log!` (the macro behand `info!` and `error!` etc) macro to retrieve and set trace id when generating logging event. It looks like this:

image

In this demo I set trace id in MySQL entry. And after that point, all the logs will contain an extra field (the italic text in the red marker) named `trace_id`. Different task/session has a different id. And if that id is not set, like on starting process, this field will not present.

The json format looks like the following, it's also available for consumers like loki
```plaintext
 176   │ {"v":0,"name":"greptime-standalone","msg":"MySQL connection coming from: 127.0.0.1:54050","level":30,"hostname":"wayne","pid":1383328,"time":"20
       │ 23-07-21T13:28:19.367622708Z","target":"servers::mysql::server","line":157,"file":"src/servers/src/mysql/server.rs"}
 177   │ {"v":0,"name":"greptime-standalone","msg":"trace_id: 123, query: select date_part('day', ts) as dt, avg(val) as avg_value from host_network_tran
       │ smit_bytes_total group by dt order by dt asc","level":30,"hostname":"wayne","pid":1383328,"time":"2023-07-21T13:28:25.105549855Z","target":"serv
       │ ers::mysql::handler","line":101,"file":"src/servers/src/mysql/handler.rs"}
 178   │ {"v":0,"name":"greptime-standalone","msg":"trace id at frontend instance: 123","level":30,"hostname":"wayne","pid":1383328,"time":"2023-07-21T13
       │ :28:25.105678985Z","target":"frontend::instance","line":490,"file":"src/frontend/src/instance.rs","trace_id":123}
 179   │ {"v":0,"name":"greptime-standalone","msg":"trace id at merge reader: None","level":30,"hostname":"wayne","pid":1383328,"time":"2023-07-21T13:28:
       │ 25.120268499Z","target":"storage::read::merge","line":504,"file":"src/storage/src/read/merge.rs"}
```

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

related to
- #2014
- #1312

### CODE REVIEW

**Comment by sunng87** [comment]
I like the idea to be able to log trace id automatically. I was thinking about adding a log function to `QueryContext` so that we can log the trace ID. But it's unfortunate that not every function has access to `QueryContext`. 

I would suggest to put `QueryContext` into the task local because there can be more context data, span ID for example, to inject. 

Second question, is this task local automatically propagate when calling `tokio::spawn`?

**Comment by waynexia** [comment]
>I would suggest to put QueryContext into the task local because there can be more context data, span ID for example, to inject.

`QueryContext` is too "huge", and contains many unrelated informations like catalog/schema. For span id, it should not be a task local variable like trace id. Each function call has its own span id, and this is planned to implement via minitrace, which will maintain its span id by macro.

>Second question, is this task local automatically propagate when calling tokio::spawn?

We don't have this kind of spawn for now. Later when the region worker is ready we need to pass trace id through region worker's request

**Comment by sunng87** [comment]
Sounds good! Thanks for the clarification.

**Review by sunng87 — APPROVED** [review]

File: src/servers/src/mysql/handler.rs [hunk 1]
```diff
@@ -97,7 +97,14 @@ impl MysqlInstanceShim {
...
         } else {
-            self.query_handler.do_query(query, query_ctx).await
+            let trace_id = query.len() as u64;
+            common_telemetry::info!("trace_id: {}, query: {}", trace_id, query);
+
+            common_telemetry::TRACE_ID
+                .scope(trace_id, async move {
+                    self.query_handler.do_query(query, query_ctx).await
+                })
+                .await
```
Comment (shuiyisong) [inline src/servers/src/mysql/handler.rs:107]: Are these test codes? 😏

Comment (waynexia) [reply src/servers/src/mysql/handler.rs]: Sure! I've changed it to the real id get from ctx

**Review by shuiyisong — APPROVED** [review]

**Pull request merged**